### PR TITLE
VER-105 Add generic v1 Harbor Dockerfile environment setup support

### PR DESCRIPTION
**Pull request by xeophon**
## Summary
- add v1 Harbor Dockerfile replay for tasks with `environment/Dockerfile` and no explicit `environment.docker_image`
- upload the Harbor `environment/` build context and translate `FROM`, `WORKDIR`, `ENV`, `RUN`, and `COPY`/`ADD` into sandbox/program setup
- export the small parser helper from the v1 taskset package

Stacked on #1392.

## Testing
- `uv run pytest tests/test_v1_harbor_cli.py -q`
- `uv run pytest tests/test_v1_mini_swe_agent.py -q`
- `uv run ruff check --fix verifiers/v1/packages/tasksets/harbor.py tests/test_v1_harbor_cli.py verifiers/v1/packages/tasksets/__init__.py verifiers/v1/__init__.py`
- `uv run ruff format --check verifiers/v1/packages/tasksets/harbor.py tests/test_v1_harbor_cli.py verifiers/v1/packages/tasksets/__init__.py verifiers/v1/__init__.py`
- `uv run ty check verifiers/v1/packages/tasksets/harbor.py`
- OpenThoughts row smoke for `broken-python`, `jq-data-processing`, `log-summary`

---

> [!NOTE]
> **Medium Risk**
> Adds new Dockerfile parsing and command-generation that affects how Harbor tasks provision their sandbox (image/workdir/env/setup), which could break task execution if parsing or path handling is wrong.
> 
> **Overview**
> **Harbor v1 tasks can now derive their runtime environment from an `environment/Dockerfile`** when `[environment].docker_image` is not set. `HarborTaskset.task_row` parses the Dockerfile to set `sandbox.image`/`sandbox.workdir`, uploads the `environment/` directory as a build context (`HARBOR_BUILD_CONTEXT`), and injects translated `ENV`/`RUN`/`COPY`/`ADD` steps into `program.env` and `program.setup`.
> 
> Introduces a public `parse_harbor_dockerfile` helper (re-exported via `verifiers.v1` / tasksets) with guardrails like rejecting multi-stage builds, and adds focused tests covering Dockerfile replay behavior and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3d73cc35ba2aee182aef7f4b3542aec80dc2efdf. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add `parse_harbor_dockerfile` to replay Dockerfile environment setup in Harbor tasks
> - Adds `parse_harbor_dockerfile` in [harbor.py](https://github.com/PrimeIntellect-ai/verifiers/pull/1407/files#diff-255f9b42c2be25db663bd7e76249d9390726ff47909524e0bb5169b6bc7611e9) that converts a single-stage Dockerfile into a replayable config: base image, workdir, ENV variables, and shell setup steps (RUN, COPY, ADD, WORKDIR).
> - Updates `HarborTaskset.task_row` to detect an `environment/Dockerfile` when no `docker_image` is configured and apply the parsed image, workdir, env, and setup steps to the sandbox and program.
> - COPY/ADD instructions are resolved relative to `HARBOR_BUILD_CONTEXT` (`/tmp/harbor_environment`), with `program.dirs` mapping the build context to the environment directory.
> - Multi-stage Dockerfiles (more than one `FROM`) are rejected with a `ValueError`.
> - Exports `parse_harbor_dockerfile` from `verifiers.v1` and `verifiers.v1.packages.tasksets`.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 3d73cc3.</sup>
> 
>

### CODE REVIEW

File: verifiers/v1/packages/tasksets/harbor.py [hunk 1]
```diff
@@ -193,12 +212,87 @@ def task_row(self, task_dir: Path, index: int) -> ConfigData:
...
+            for source in paths[:-1]:
+                source = source[2:] if source.startswith("./") else source.lstrip("/")
+                source_path = (
+                    build_context
+                    if source in {"", "."}
+                    else f"{build_context}/{source}"
+                )
+                setup.append(
+                    f"mkdir -p {shlex.quote(mkdir_path)} && "
+                    f"cp -R {shlex.quote(source_path)} {shlex.quote(target)}"
```
[the remark this answers is not in the document]
Comment (xeophon) [reply verifiers/v1/packages/tasksets/harbor.py:303]: Addressed: COPY sources that are the build context, end with a slash, or resolve to a local directory now copy contents via source/. so COPY . . no longer nests /tmp/harbor_environment under the workdir. Added a regression test for COPY . .

File: verifiers/v1/packages/tasksets/harbor.py [hunk 2]
```diff
@@ -193,12 +212,87 @@ def task_row(self, task_dir: Path, index: int) -> ConfigData:
...
+        parts = line.split(None, 1)
+        if len(parts) < 2:
+            continue
+        kind, value = parts[0].upper(), parts[1].strip()
+        if kind == "FROM":
+            image = next(
+                token
+                for token in value.split()
+                if not token.startswith("--") and token.upper() != "AS"
+            )
```
[the remark this answers is not in the document]
Comment (xeophon) [reply verifiers/v1/packages/tasksets/harbor.py:257]: Addressed: multi-stage Dockerfiles now fail during HarborTaskset Dockerfile replay with a clear ValueError telling users to provide a prebuilt [environment].docker_image instead. Added a regression test for this path.

File: verifiers/v1/packages/tasksets/harbor.py [hunk 3]
```diff
@@ -193,12 +212,87 @@ def task_row(self, task_dir: Path, index: int) -> ConfigData:
...
+
+    for raw_line in dockerfile_path.read_text().splitlines():
+        line = raw_line.rstrip()
+        if line.endswith("\\"):
+            pending += line[:-1].rstrip() + " "
+            continue
+        line = (pending + line).strip()
+        pending = ""
+        if not line or line.startswith("#"):
+            continue
```
[the remark this answers is not in the document]
Comment (xeophon) [reply verifiers/v1/packages/tasksets/harbor.py:240]: Addressed: standalone blank/comment lines are now skipped before line-continuation handling, so a comment ending in backslash no longer swallows the next Dockerfile instruction. Added a regression test.

File: verifiers/v1/packages/tasksets/harbor.py [hunk 4]
```diff
@@ -193,12 +212,87 @@ def task_row(self, task_dir: Path, index: int) -> ConfigData:
...
+            setup.append(f"mkdir -p {shlex.quote(workdir)}")
+        elif kind == "ENV":
+            tokens = shlex.split(value)
+            if tokens and "=" in tokens[0]:
+                for token in tokens:
+                    if "=" in token:
+                        key, _, val = token.partition("=")
+                        env[key] = val
+            elif len(tokens) >= 2:
+                env[tokens[0]] = tokens[1]
```
[the remark this answers is not in the document]
Comment (xeophon) [reply verifiers/v1/packages/tasksets/harbor.py]: Addressed: legacy ENV key value form now preserves all value tokens by joining tokens[1:], so ENV GREETING hello beautiful world becomes GREETING="hello beautiful world". Added a regression test.
